build.sh

```diff
@@ -446,7 +446,11 @@ build_CoreLib()
     if [[ "$__BuildManagedTools" -eq "1" ]]; then
         echo "Publishing crossgen2 for $__DistroRid"
         "$__ProjectRoot/dotnet.sh" publish --self-contained -r $__DistroRid -c $__BuildType -o "$__BinDir/crossgen2" "$__ProjectRoot/src/tools/crossgen2/crossgen2/crossgen2.csproj"
+
+        # Attempt to copy both .so and .dylib to cover Mac and other unixes
```

Comment:
I would prefer using the correct extension instead of trying which one we hit. You can compare 
the `__HostOS` env var to "OSX". 
However, I wonder why do we put the crossgen2 into a separate subfolder and use a renamed copies of the jit / jitinterface. All other binaries we build (including old crossgen) are going into the same flat folder. Why does crossgen2 have to be different?

Reply:
> I would prefer using the correct extension instead of trying which one we hit

I preferred less code to nice code, but I'll fix that.

> However, I wonder why do we put the crossgen2 into a separate subfolder

It's a self-contained publish and has it's own coreclr.dll, etc. crossgen2 is a managed app and we need to be able to launch it. CORE_ROOT doesn't have the framework in it. It was decided we would run it on a known good runtime/framework for now instead of trying bleeding edge of everything.

Reply:
I am not convinced that this is a good reason for putting it into a separate folder. For example, R2RDump binaries or runincontext tool binaries are stored into the same folder and it kind of matches what you've said about crossgen2 too. The `__BinDir` folder is a flat drop folder for all the artefacts that coreclr native and managed build produces. It is true that there are couple of subdirectories (.nuget, PDB, inc, lib, IL, gcinfo), but these are either special purpose or helper folders.

Reply:
We need something that we can execute during testing and eventually: execute to crossgen CoreLib as part of the CoreCLR build.

Runincontext kind of just injects itself before an existing test and we corerun runincontext instead of the test. It doesn't do anything interesting and it doesn't have any dependencies. R2rdump we don't really test so it isn't interesting. So these don't really map to what we need for crossgen2.

Crossgen2 is a tool we need to be able to execute as a standalone tool before a test executes. It depends on NuGet packages we probably don't want in CORE_ROOT.

There might also be chicken and egg in the future: we use crossgen to generate the System.Private.CoreLib that is in BinDir. Once crossgen is replaced by crossgen2, how would we launch crossgen2 out of BinDir if CoreLib is still not there, etc.

I kind of wish we would just compile crossgen2 with the CoreRT compiler and then we can dump it anywhere we want because it's just a native tool with no dependencies like anything else in BinDir.

If you have any ideas how we can make this work, I would be happy to change it to something else.

Reply:
I think we can run crossgen2 using the dotnet in .dotnet/dotnet folder that the arcade installs there for managed build purposes. Since that one needs to be there to actually build crossgen2, there doesn't seem to be a chicken and egg problem even if we started to use crossgen2 instead of crossgen.

However, I have looked at the crossgen2 folder and I can see now that there is a plenty of managed files in there, various System.XXX files. I hope this we could rid of those and just use the .dotnet/dotnet. Then we could have crossgen2 right in the BinDir. You are right that in the current state, it would not be a good idea to put all of those to the BinDir though.

Reply:
Btw, I've just found that if you change the TargetFramework in crossgen2.csproj to netcoreapp3.0, then it builds correctly on Alpine too. The .dotnet/dotnet tool is 3.0.100-preview6-012264.

Reply:
Ok, let's keep it the way you have it. I understand the desire for the published crossgen2 now.